### ISSUE TITLE
How to export all symbols of a shared library on Windows by default?

### ISSUE THREAD

**Issue by lyrahgames** [build2/build2#315]
I try to create a [build2 package](https://github.com/build2-packaging/geometry-central) for the library [Geometry Central](https://github.com/nmwsharp/geometry-central). Unfortunately, the library does not use any export header functionality for the exportation of symbols when building a shared library. Instead, it relies on a [CMake feature](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) to export all symbols on Windows by default. A few insights on that might be given [here](https://stackoverflow.com/questions/225432/export-all-symbols-when-creating-a-dll). As a consequence, the build2 package version cannot be built on Windows machines as MSVC doesn't see any exported symbols. Is there a way in build2 to do the same or achieve something similar? Otherwise, do you have any suggestions what I could do to fix [the issue](https://github.com/build2-packaging/geometry-central/issues/1)?

**Comment by Klaim** [comment]
Not completely sure if it's exactly equivalent to how CMakes does it but there is a (limited) way with `build2`: https://build2.org/build2/doc/build2-build-system-manual.xhtml#cc-auto-symexport

**Comment by lyrahgames** [comment]
I have added the automatic DLL symbol exportation according to the [build2 manual](https://build2.org/build2/doc/build2-build-system-manual.xhtml#cc-auto-symexport). The current CI run can be found [here](https://ci.cppget.org/@9b308b44-7877-4171-a384-2b5883b939e6). Unfortunately, it still does not work for MSVC on Windows. Now, the library is compiled, though. It only fails the tests with a few errors like the following.
```
error LNK2019: unresolved external symbol "const geometrycentral::surface::SurfaceMesh::`vftable'"
```
According to the manual:
> Note also that this only deals with exporting of the symbols from a DLL. In order to work, code that uses such a DLL should be able to import the symbols without explicit __declspec(dllimport) declarations. This works thanks to the symbol auto-importing support in Windows linkers. Note, however, that auto-importing only works for functions and not for global variables.

This might be a problem of the library itself because the `vftable` used for polymorphism might count as a global variable. Maybe you have an idea to fix it. Otherwise, I will close this issue as I was able to successfully export all functions by default.

**Comment by boris-kolpackov** [comment]
If this works with CMake then it should also work with `build2` since both implementations use essentially the same approach. We may be filtering out something we shouldn't (our implementation tries to be quite a bit "tighter" by not exporting unnecessary stuff).

If you can create a minimal reproducer, I would take a look.

**Comment by boris-kolpackov** [comment]
> Currently, I do not have access to a Windows machine.

You could probably get away with the just using CI. I.e., create a test project (for example, in your personal GitHub) with the relevant code and CI it until you get the same error. Feels like you should be able to narrow it down sufficiently in a couple of runs.

**Comment by lyrahgames** [comment]
I was able to flesh out a minimal reproducer. You can find the repository [here](https://github.com/lyrahgames/build2-export-test). I have added some comments to mark the essential parts. Unsurprisingly, you can find the main stuff in the following files.
- `libmytest/buildfile`
- `libmytest/mytest.hpp`
- `libmytest/mytest.cpp`
- `tests/basics/driver.cpp`

[Here](https://ci.cppget.org/@0fc1e96d-1b7d-41b2-9c4a-934d1ea4907d) is the latest CI run. Except three MSVC-based builds, all builds are successful. The MSVC-based builds emit the following error:

```
error LNK2019: unresolved external symbol "const mytest::bug::`vftable'"
referenced in function "public: __cdecl mytest::bug::bug(struct mytest::wrapper const &)"
```

In short, the error is generated by a call to an inlined constructor of a class with virtual function table that calls an alternative constructor defined in the source files.

**Comment by boris-kolpackov** [comment]
Thanks for the reproducer (I hope you don't mind I've copied the relevant files to our internal test for this functionality)!

Turns out we do export the relevant symbol but we mark it as `DATA` in the `.def` file because it comes from the `rdata` (read-only data) section. But it looks like this is an incorrect assumption since removing `DATA` for this symbols allows everything to long successfully. So I've removed `DATA` from all `rdata` symbols, at least for MSVC (GCC doesn't seem to be affected by that, though their `.def` format supports `CONSTANT` which could be more appropriate).

This fix will be available in the tomorrow's staged toolchain.

**Comment by boris-kolpackov** [comment]
Ok, the fix is staged and here is a full CI run of our test: https://ci.stage.build2.org/@ff3e2aa9-8a01-4c74-8b08-1b45a3e2c1d8 You can also test your original library using the ci.stage.build2.org service. I am going to close this but feels free to reopen/comment if there are still any issues.

**Comment by lyrahgames** [comment]
> Thanks for the reproducer (I hope you don't mind I've copied the relevant files to our internal test for this functionality)!

And thank you for the quick fix! No, I don't mind at all.

> This fix will be available in the tomorrow's staged toolchain.

> Ok, the fix is staged and here is a full CI run of our test: https://ci.stage.build2.org/@ff3e2aa9-8a01-4c74-8b08-1b45a3e2c1d8 You can also test your original library using the ci.stage.build2.org service. I am going to close this but feels free to reopen/comment if there are still any issues.

I have installed the staged toolchain, tested it with the original library, and can confirm that the error is fixed.
